test/e2e/resource_usage_gatherer.go

```diff
@@ -75,43 +70,193 @@ func (s *ResourceUsageSummary) PrintJSON() string {
+}
+
+func (w *resourceGatherWorker) gather(initialSleep time.Duration, errors []error) {
+	defer utilruntime.HandleCrash()
+	defer w.wg.Done()
+	select {
+	case <-time.After(initialSleep):
+		w.singleProbe(errors)
 		for {
 			select {
-			case <-time.After(period):
-				now := time.Now()
-				data, err := g.getKubeSystemContainersResourceUsage(c)
-				if err != nil {
-					Logf("Error while getting resource usage: %v", err)
-					continue
-				}
-				g.usageTimeseries[now] = data
-			case <-g.stopCh:
-				Logf("Stop channel is closed. Stopping gatherer.")
-				return nil
+			case <-time.After(resourceDataGatheringPeriod):
+				w.singleProbe(errors)
+			case <-w.stopCh:
+				return
 			}
 		}
+	case <-w.stopCh:
+		return
+	}
+}
+
+func (g *containerResourceGatherer) getKubeSystemContainersResourceUsage(c *client.Client) error {
+	errors := make([]error, 0)
+	delay := resourceDataGatheringPeriod / time.Duration(len(g.workers))
+	go func() {
+		for i := range g.workers {
+			go g.workers[i].gather(delay, errors)
+		}
 	}()
+	g.workerWg.Wait()
+
+	if len(errors) != 0 {
+		return fmt.Errorf("Errors while gathering usage data: %v", errors)
+	}
+	return nil
+}
+
+type containerResourceGatherer struct {
+	client               *client.Client
+	stopCh               chan struct{}
+	workers              []resourceGatherWorker
+	workerWg             sync.WaitGroup
+	containerIDToNameMap map[string]string
+	containerIDs         []string
+}
+
+func NewResourceUsageGatherer(c *client.Client) *containerResourceGatherer {
+	g := containerResourceGatherer{
+		client:               c,
+		stopCh:               make(chan struct{}),
+		containerIDToNameMap: make(map[string]string),
+		containerIDs:         make([]string, 0),
+	}
+
+	pods, err := c.Pods("kube-system").List(api.ListOptions{})
+	if err != nil {
+		Logf("Error while listing Pods: %v", err)
+		return nil
+	}
+	for _, pod := range pods.Items {
+		for _, container := range pod.Status.ContainerStatuses {
+			containerID := strings.TrimPrefix(container.ContainerID, "docker:/")
+			g.containerIDToNameMap[containerID] = pod.Name + "/" + container.Name
+			g.containerIDs = append(g.containerIDs, containerID)
+		}
+	}
+	var nodeList *api.NodeList
+	if wait.PollImmediate(poll, 10*time.Second, func() (bool, error) {
+		nodeList, err = c.Nodes().List(api.ListOptions{})
+		return err == nil, nil
+	}) != nil {
+		expectNoError(err, "Timed out while listing nodes for e2e cluster.")
+	}
+
+	g.workerWg.Add(len(nodeList.Items))
+	for _, node := range nodeList.Items {
+		g.workers = append(g.workers, resourceGatherWorker{
+			c:                    c,
+			nodeName:             node.Name,
+			wg:                   &g.workerWg,
+			containerIDToNameMap: g.containerIDToNameMap,
+			containerIDs:         g.containerIDs,
+			stopCh:               g.stopCh,
+		})
+	}
+	return &g
+}
+
+func (g *containerResourceGatherer) startGatheringData() {
+	go g.getKubeSystemContainersResourceUsage(g.client)
```

Comment:
maybe simply paste the content of "getKubeSystemContainersRU" here and say that startGatheringData() is blocking and run it with go